src/Sylius/Bundle/ApiBundle/Serializer/ProductVariantSerializer.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Serializer;
+
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Core\Calculator\ProductVariantPriceCalculatorInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Webmozart\Assert\Assert;
+
+final class ProductVariantSerializer implements NormalizerInterface
+{
+    /** @var NormalizerInterface */
+    private $objectNormalizer;
+
+    /** @var ProductVariantPriceCalculatorInterface */
+    private $priceCalculator;
+
+    /** @var ChannelContextInterface */
+    private $channelContext;
+
+    public function __construct(
+        NormalizerInterface $objectNormalizer,
+        ProductVariantPriceCalculatorInterface $priceCalculator,
+        ChannelContextInterface $channelContext
+    ) {
+        $this->objectNormalizer = $objectNormalizer;
+        $this->priceCalculator = $priceCalculator;
+        $this->channelContext = $channelContext;
+    }
+
+    public function normalize($object, $format = null, array $context = [])
+    {
+        Assert::isInstanceOf($object, ProductVariantInterface::class);
+
+        $data = $this->objectNormalizer->normalize($object, $format, $context);
+        if (isset($data['product'])) {
+            $data['product'] = $data['product']['@id'];
+            $data['translations'] = $this->serializeTranslation($data['translations']);
+        }
+        $data['price'] = $this->priceCalculator->calculate($object, ['channel' => $this->channelContext->getChannel()]);
+
+        return $data;
+    }
+
+    public function supportsNormalization($data, $format = null, $context = []): bool
+    {
+        return
+            $data instanceof ProductVariantInterface &&
+            $context['operation_type'] !== 'subresource' &&
+            $context['groups'] != 'shop:cart:read' &&
+            $context['groups'] != 'admin:order_item_unit:read';
```

Comment:
Maybe you should allow for all shop routes?

Reply:
im still wokring on it, gona refactor before finishing, my goal is to get all tests green